File: build_gazetteer.py
#!/usr/bin/env python3
"""Build a compact JSON gazetteer for in-browser geocoding.

Layout (gazetteer.json):
{
  "countries": [{"cc": "US", "name": "United States", "lat":..., "lon":...}, ...],
  "admin1":    [{"cc": "US", "code": "VA", "name": "Virginia", "lat":..., "lon":...}, ...],
  "us_counties": [{"st":"VA","name":"henrico","lat":...,"lon":...}, ...],
  "cities":   compact array: alternating arrays per row,
              [name_slug, cc, admin1_code, lat, lon, pop]
}
"""
import json
import re
import sys
import unicodedata
import zipfile
from collections import defaultdict
from pathlib import Path

ROOT = Path(__file__).parent
MIN_POP = int(sys.argv[1]) if len(sys.argv) > 1 else 1000


def slug(s: str) -> str:
    s = unicodedata.normalize("NFKD", s).encode("ascii", "ignore").decode("ascii")
    return re.sub(r"[^a-z0-9]+", " ", s.lower()).strip()


def main() -> None:
    countries = []
    cc_to_name = {}
    with (ROOT / "countryInfo.txt").open() as f:
        for line in f:
            if line.startswith("#") or not line.strip():
                continue
            parts = line.rstrip("\n").split("\t")
            if len(parts) < 9:
                continue
            iso, _iso3, _isoNum, _fips, name = parts[0], parts[1], parts[2], parts[3], parts[4]
            cc_to_name[iso] = name

    admin1 = []
    admin1_keys = []
    with (ROOT / "admin1.txt").open() as f:
        for line in f:
            parts = line.rstrip("\n").split("\t")
            if len(parts) < 4:
                continue
            code, name, ascii_name, _gid = parts
            cc, st = code.split(".", 1)
            admin1.append({"cc": cc, "code": st, "name": ascii_name or name})
            admin1_keys.append((cc, st))

    us_counties = []
    with zipfile.ZipFile(ROOT / "US.zip") as z:
        with z.open("US.txt") as f:
            for raw in f:
                p = raw.decode("utf-8", "replace").rstrip("\n").split("\t")
                if len(p) < 15 or p[6] != "A" or p[7] != "ADM2":
                    continue
                ascii_name = p[2]
                try:
                    lat = float(p[4]); lon = float(p[5])
                except ValueError:
                    continue
                st = p[10]
                key = slug(ascii_name).replace(" county", "").replace(" parish", "").strip()
                us_counties.append({
                    "st": st,
                    "name": key,
                    "lat": round(lat, 4),
                    "lon": round(lon, 4),
                })

    cities = []
    state_lat_acc: dict[tuple[str, str], list[float]] = defaultdict(lambda: [0.0, 0.0, 0])
    country_lat_acc: dict[str, list[float]] = defaultdict(lambda: [0.0, 0.0, 0])
    with zipfile.ZipFile(ROOT / "cities500.zip") as z:
        with z.open("cities500.txt") as f:
            for raw in f:
                p = raw.decode("utf-8", "replace").rstrip("\n").split("\t")
                if len(p) < 15:
                    continue
                if p[6] != "P":
                    continue
                try:
                    lat = float(p[4]); lon = float(p[5])
                    pop = int(p[14])
                except ValueError:
                    continue
                cc = p[8]
                st = p[10]
                w = max(pop, 1)
                state_lat_acc[(cc, st)][0] += lat * w
                state_lat_acc[(cc, st)][1] += lon * w
                state_lat_acc[(cc, st)][2] += w
                country_lat_acc[cc][0] += lat * w
                country_lat_acc[cc][1] += lon * w
                country_lat_acc[cc][2] += w
                if pop < MIN_POP:
                    continue
                ascii_name = p[2] or p[1]
                alt = p[3]
                cities.append([
                    slug(ascii_name), cc, st,
                    round(lat, 4), round(lon, 4), pop,
                ])

    for a in admin1:
        key = (a["cc"], a["code"])
        if state_lat_acc[key][2] > 0:
            tw = state_lat_acc[key][2]
            a["lat"] = round(state_lat_acc[key][0] / tw, 4)
            a["lon"] = round(state_lat_acc[key][1] / tw, 4)
        else:
            a["lat"] = 0.0
            a["lon"] = 0.0

    for cc, name in cc_to_name.items():
        if country_lat_acc[cc][2] > 0:
            tw = country_lat_acc[cc][2]
            countries.append({
                "cc": cc, "name": name,
                "lat": round(country_lat_acc[cc][0] / tw, 4),
                "lon": round(country_lat_acc[cc][1] / tw, 4),
            })

    out = {
        "countries": countries,
        "admin1": admin1,
        "us_counties": us_counties,
        "cities": cities,
    }
    out_path = ROOT / "gazetteer.json"
    out_path.write_text(json.dumps(out, separators=(",", ":")))
    print(f"countries: {len(countries)}")
    print(f"admin1:    {len(admin1)}")
    print(f"counties:  {len(us_counties)}")
    print(f"cities:    {len(cities)} (pop >= {MIN_POP})")
    print(f"size:      {out_path.stat().st_size / 1024 / 1024:.1f} MB")


if __name__ == "__main__":
    main()

File: src/bin/build-gazetteer.ts
#!/usr/bin/env -S npx tsx
import { readFileSync, writeFileSync, statSync } from "node:fs";
import { argv, exit } from "node:process";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import AdmZip from "adm-zip";
import { slug } from "../geo/slug.js";

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = join(HERE, "..", "..");

const minPop = Number.parseInt(argv[2] ?? "1000", 10);

interface Country { cc: string; name: string; lat: number; lon: number; }
interface Admin1 { cc: string; code: string; name: string; lat: number; lon: number; }
interface UsCounty { st: string; name: string; lat: number; lon: number; }
type CityRow = [string, string, string, number, number, number];

function readGeonamesText(zipPath: string, innerName: string): string {
  const zip = new AdmZip(zipPath);
  const entry = zip.getEntry(innerName);
  if (!entry) throw new Error(`${innerName} not found in ${zipPath}`);
  return entry.getData().toString("utf8");
}

function round4(n: number): number { return Math.round(n * 10000) / 10000; }

function main(): number {
  const ccToName = new Map<string, string>();
  for (const line of readFileSync(join(ROOT, "countryInfo.txt"), "utf8").split(/\r?\n/)) {
    if (!line || line.startsWith("#")) continue;
    const parts = line.split("\t");
    if (parts.length < 9) continue;
    const iso = parts[0]!;
    const name = parts[4]!;
    ccToName.set(iso, name);
  }

  const admin1: Admin1[] = [];
  for (const line of readFileSync(join(ROOT, "admin1.txt"), "utf8").split(/\r?\n/)) {
    if (!line) continue;
    const parts = line.split("\t");
    if (parts.length < 4) continue;
    const code = parts[0]!;
    const name = parts[1]!;
    const asciiName = parts[2]!;
    const dot = code.indexOf(".");
    if (dot < 0) continue;
    admin1.push({
      cc: code.slice(0, dot),
      code: code.slice(dot + 1),
      name: asciiName || name,
      lat: 0, lon: 0,
    });
  }

  const usCounties: UsCounty[] = [];
  const usText = readGeonamesText(join(ROOT, "US.zip"), "US.txt");
  for (const line of usText.split(/\r?\n/)) {
    if (!line) continue;
    const p = line.split("\t");
    if (p.length < 15 || p[6] !== "A" || p[7] !== "ADM2") continue;
    const asciiName = p[2]!;
    const lat = Number.parseFloat(p[4]!);
    const lon = Number.parseFloat(p[5]!);
    if (!Number.isFinite(lat) || !Number.isFinite(lon)) continue;
    const st = p[10]!;
    const key = slug(asciiName).replace(" county", "").replace(" parish", "").trim();
    usCounties.push({ st, name: key, lat: round4(lat), lon: round4(lon) });
  }

  const cities: CityRow[] = [];
  const stateAcc = new Map<string, [number, number, number]>();
  const countryAcc = new Map<string, [number, number, number]>();
  const accKey = (cc: string, st: string) => `${cc}|${st}`;

  const citiesText = readGeonamesText(join(ROOT, "cities500.zip"), "cities500.txt");
  for (const line of citiesText.split(/\r?\n/)) {
    if (!line) continue;
    const p = line.split("\t");
    if (p.length < 15 || p[6] !== "P") continue;
    const lat = Number.parseFloat(p[4]!);
    const lon = Number.parseFloat(p[5]!);
    const pop = Number.parseInt(p[14]!, 10);
    if (!Number.isFinite(lat) || !Number.isFinite(lon) || !Number.isFinite(pop)) continue;
    const cc = p[8]!;
    const st = p[10]!;
    const w = Math.max(pop, 1);
    const sk = accKey(cc, st);
    const sa = stateAcc.get(sk) ?? [0, 0, 0];
    sa[0] += lat * w; sa[1] += lon * w; sa[2] += w;
    stateAcc.set(sk, sa);
    const ca = countryAcc.get(cc) ?? [0, 0, 0];
    ca[0] += lat * w; ca[1] += lon * w; ca[2] += w;
    countryAcc.set(cc, ca);
    if (pop < minPop) continue;
    const asciiName = p[2] || p[1]!;
    cities.push([slug(asciiName), cc, st, round4(lat), round4(lon), pop]);
  }

  for (const a of admin1) {
    const acc = stateAcc.get(accKey(a.cc, a.code));
    if (acc && acc[2] > 0) {
      a.lat = round4(acc[0] / acc[2]);
      a.lon = round4(acc[1] / acc[2]);
    }
  }

  const countries: Country[] = [];
  for (const [cc, name] of ccToName) {
    const acc = countryAcc.get(cc);
    if (acc && acc[2] > 0) {
      countries.push({
        cc, name,
        lat: round4(acc[0] / acc[2]),
        lon: round4(acc[1] / acc[2]),
      });
    }
  }

  const out = { countries, admin1, us_counties: usCounties, cities };
  const outPath = join(ROOT, "gazetteer.json");
  writeFileSync(outPath, JSON.stringify(out));
  const sizeMb = (statSync(outPath).size / 1024 / 1024).toFixed(1);
  console.log(`countries: ${countries.length}`);
  console.log(`admin1:    ${admin1.length}`);
  console.log(`counties:  ${usCounties.length}`);
  console.log(`cities:    ${cities.length} (pop >= ${minPop})`);
  console.log(`size:      ${sizeMb} MB`);
  return 0;
}

exit(main());
